### PR TITLE
[8.x] Add binding field to explicit model binding

### PR DESCRIPTION
**Pull request by amir9480**
This PR adds binding field support to model bindings registered via `Route::model`.

Example:

RouteSerivceProvider.php:
```php
Route::model('root_category', Category::class);
```

routes/web.php:
```php
Route::get('category/{root_category:slug}', CategoryController::class)->name('category');
```

Using `route` function will give us this output:
```php
route('category', ['root_category' => $category]); // http://localhost/category/category-slug
```

But opening the URL shows a `404` error page.

This PR fixes this issue.

(Also test for [parent-child](https://github.com/laravel/framework/pull/36966#issuecomment-818705838) scope added.)

Edit:
It's also possible to use the binding field for closure bind.

Example:
```php
Route::bind('custom_value', function ($value, $route = null, $field = null) {
    return Model::where($field ?? 'id', $value)->firstOrFail();
});
```

### CODE REVIEW

**Comment by driesvints** [comment]
Your PR already got rejected here: https://github.com/laravel/framework/pull/36966

**Comment by amir9480** [comment]
@driesvints 
image

**Comment by driesvints** [comment]
Sorry! Missed that

**Comment by taylorotwell** [comment]
Where are you actually testing the query was scoped?

**Comment by amir9480** [comment]
@taylorotwell 
I'm not sure about tests but at
[RoutingRouteTest.php:1697](https://github.com/amir9480/framework/blob/29073c4c25ccbc53f11075a64afefbc78fd34408/tests/Routing/RoutingRouteTest.php#L1697)

I write this test based on the test for implicit binding ([RoutingRouteTest.php:1663](https://github.com/amir9480/framework/blob/29073c4c25ccbc53f11075a64afefbc78fd34408/tests/Routing/RoutingRouteTest.php#L1663))

If there are more tests for parent-child implicit binding please let me know, so I can write tests for explicit binding based on them.

**Comment by amir9480** [comment]
@taylorotwell @driesvints
Now I think understood how parent-child works and I didn't know I should change `forModel` implementation.

Now there is a problem with implementing this feature
As explicit bindings work in separate loop iteration before implicit binding, if we put an explicit binding model after implicit it won't work.

[SubstituteBindings.php:39](https://github.com/amir9480/framework/blob/29073c4c25ccbc53f11075a64afefbc78fd34408/src/Illuminate/Routing/Middleware/SubstituteBindings.php#L39)

Example:
```php
Route::model('custom_post', Post::class);

Route::get('/user/{user}/{custom_post:slug}', ...); // Won't work
```

The issue here in this case user parameter should be initialized before calling `custom_post` binding. but I'm not sure how I should change the current implementation

**Comment by driesvints** [comment]
@amir9480 Taylor won't see this until the pr is out of draft.

File: src/Illuminate/Routing/RouteBinding.php [hunk 1]
```diff
@@ -67,7 +70,23 @@ public static function forModel($container, $class, $callback = null)
...
-            if ($model = $instance->resolveRouteBinding($value)) {
+            $parent = $route->parentOfParameter($key);
+
+            if ($parent instanceof UrlRoutable && in_array($key, array_keys($route->bindingFields()))) {
+                try {
+                    $model = $parent->resolveChildRouteBinding(
+                        $key, $value, $field
+                    );
+                } catch (BadMethodCallException $exception) {
+                    $model = $parent->resolveChildRouteBinding(
```
Comment (amir9480) [inline src/Illuminate/Routing/RouteBinding.php:81]: This is using for the undefined explicit model binding name.

For example:
```php
Route::model('special_post', Post::class);

Route::get('foo/{user}/{special_post:slug}');
```

At first, it looks for the `specialPosts` relationship in the `user` model, If `specialPosts` does not exist then it's gonna use `posts` relationship.

If you think it's a Unnecessary behavior let me know.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Pull request closed without merging**